makeWrapper: quote variables

abbradar opened the pull request:
###### Motivation for this change

Fix things like `--prefix FOO ' ' 'bar baz'`

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

orivej commented:
Could you post an example of the resulting wrapper before and after this change?

abbradar commented:
Before:

```
export NIX_CFLAGS_COMPILE=-isystem /nix/store/3cv0q9g9zyg5alzzz6072rhdmvzhk5yy-cuda-floatn.h/include${NIX_CFLAGS_COMPILE:+ }$NIX_CFLAGS_COMPILE
```

After:

```
export NIX_CFLAGS_COMPILE="-isystem /nix/store/3cv0q9g9zyg5alzzz6072rhdmvzhk5yy-cuda-floatn.h/include${NIX_CFLAGS_COMPILE:+ }$NIX_CFLAGS_COMPILE"
```

orivej commented:
This swaps no quoting for approximately correct quoting. Please do not add double quotes, use `${value@Q}` and `${separator@Q}` instead. The result will look like this:
```
export NIX_CFLAGS_COMPILE='-isystem /nix/store/3cv0q9g9zyg5alzzz6072rhdmvzhk5yy-cuda-floatn.h/include'${NIX_CFLAGS_COMPILE:+' '}$NIX_CFLAGS_COMPILE
```

abbradar committed:
makeWrapper: quote values properly

This uses Bash ${foo@Q} feature to quote values properly, which allows us to
handle values containing spaces, dollars etc.

Thanks orivej for the idea!

abbradar commented:
Excellent, thank you for the catch! To be honest I didn't know about @Q Bash feature before, it makes many things which I have in mind simpler.

Can you review this version? It passes my new [test](https://gist.github.com/abbradar/4233f2a8cf1a4747fe51160494d44cd7) for me with the following wrapper:

```
export FOO='ahaha $ \'
export BAR='ahaha $ \'
export BAZ='ahaha $ \'
export QUZ='ahaha $ \'
export FOO='bar '\''$baz'${FOO:+' '}$FOO
export BAR=$BAR${BAR:+'' ''}'bar '\''$baz'
export BAZ=$'test1\n test3\ntest4'${BAZ:+$}$BAZ
export QUZ=$QUZ${QUZ:+$}$'test1\n test3\ntest4'
```

abbradar commented:
@grahamc Bot forgot to remove old labels after a force-push.

abbradar commented:
How can I quote `@Q` in a commit message so that to not ping `@Q` the member with it? @Q, sorry for the noise!

orivej commented on pkgs/build-support/setup-hooks/make-wrapper.sh:
```diff
@@ -63,9 +63,9 @@ makeWrapper() {
             if test -n "$value"; then
                 if test "$p" = "--suffix"; then
-                    echo "export $varName=\$$varName\${$varName:+$separator}$value" >> "$wrapper"
+                    echo "export $varName=\$$varName\${$varName:+'${separator@Q}'}${value@Q}" >> "$wrapper"
```
These `'` should not be added.

abbradar replied:
Oops, leftover. Fixed.

orivej commented:
> To be honest I didn't know about `@Q` Bash feature before, it makes many things which I have in mind simpler.

I've learnt about it not long ago; before that I'd have suggested to use `$(printf %q "$var")` which achieves the same a bit more verbosely.

orivej reviewed (approved):
Now this looks good. (But I've not tested it.)

orivej commented:
I have set up a Hydra cluster and tested this. Thank you!

orivej commented:
Somehow I missed some failures… This is causing a bit of trouble in `staging` in cases such as [`--set CC "$""{CC:-$CC""}"`](https://github.com/NixOS/nixpkgs/blob/fa2bf1bf200218de2a894088c5a143c961663650/pkgs/development/compilers/dmd/default.nix#L155) and [`--set RETROFE_PATH "\${RETROFE_PATH:-\$PWD}"`](https://github.com/NixOS/nixpkgs/blob/fa2bf1bf200218de2a894088c5a143c961663650/pkgs/misc/emulators/retrofe/default.nix#L68). I'm going to introduce `--set-default` for the first kind of usage (set to a constant unless already set) and `--set-eval` for the latter (set to an arbitrary expression).
/cc @vcunat

vcunat commented:
[The docs](https://nixos.org/nixpkgs/manual/#ssec-stdenv-functions) refer to the source files for such details, so I see no direct need to update the manual.  It seems too low-level detail for release notes, too, not affecting any "normal" uses of the wrappers.

orivej commented:
I've reviewed all usages of makeWrapper / wrapProgram in Nixpkgs and found 6 additional expressions that had to be updated in fb703ad1173abe80740e846b10e448d7c7f1d87c (and eb3d20761169696f78bfd40afb92ab099c7507a3). It is not unlikely that when this reaches master there will be a couple runtime breakages invisible on Hydra and missed during the manual review.

Infinisil commented:
Well that explains why a few of my wrappers suddenly stopped working..
